cmd/nerdctl/container_logs.go

```diff
@@ -0,0 +1,114 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/cmd/container"
+	"github.com/spf13/cobra"
+)
+
+func newLogsCommand() *cobra.Command {
+	var logsCommand = &cobra.Command{
+		Use:               "logs [flags] CONTAINER",
+		Args:              IsExactArgs(1),
+		Short:             "Fetch the logs of a container. Currently, only containers created with `nerdctl run -d` are supported.",
+		RunE:              logsAction,
+		ValidArgsFunction: logsShellComplete,
+		SilenceUsage:      true,
+		SilenceErrors:     true,
+	}
+	logsCommand.Flags().BoolP("follow", "f", false, "Follow log output")
+	logsCommand.Flags().BoolP("timestamps", "t", false, "Show timestamps")
+	logsCommand.Flags().StringP("tail", "n", "all", "Number of lines to show from the end of the logs")
+	logsCommand.Flags().String("since", "", "Show logs since timestamp (e.g. 2013-01-02T13:23:37Z) or relative (e.g. 42m for 42 minutes)")
+	logsCommand.Flags().String("until", "", "Show logs before a timestamp (e.g. 2013-01-02T13:23:37Z) or relative (e.g. 42m for 42 minutes)")
+	return logsCommand
+}
+
+func processContainerLogsCommandOptions(cmd *cobra.Command) (types.ContainerLogsCommandOptions, error) {
```

Comment:
We may need to change `ContainerLogsCommandOptions` to `ContainerLogsOptions` (e.g. from `${Command}CommandOptions` to `${Command}Options`) for brevity. And also put `stdin/stdout/stderr` in `${Command}Options` directly (if they're used).

https://github.com/containerd/nerdctl/issues/1889 has more discussion, but the tl;dr is:

1. Rename `${Command}CommandOptions` to `${Command}Options`;
2. Put necessary `io` used by the command directly in `${Command}Options`;
3. Argument order: `func Command(ctx context.Context, param1, param2 type, options ${Command}Options` (`param1, param2 type` means, e.g., container_id, image_ref, id, etc).

Sorry to back and forth on this, but now I think it brings more consistency :) lmk if there is something unclear.

Reply:
https://github.com/containerd/nerdctl/pull/1900 has changes following up the issue.

Reply:
Thank you for the very clear summary! Updated accordingly.

Re. https://github.com/containerd/nerdctl/issues/1889#issuecomment-1397853958:

> These are optional, and nil values should imply os.Stdin, os.Stdout, os.Stderr .

I didn't add a check for this as they could be `nil` even before they are put into the options struct, and we weren't doing the checks. Checked #1900, and it seems that it's also not adding this check. Last, currently every caller in `cmd/nerdctl` should just explicitly specify them anyway.